### PR TITLE
Android 6 Support

### PR DESCRIPTION
**Pull request by M-Wong**
Use EncryptedSharedPreferences on Android 6 for transfer code keypairs and fix some tint and coloring issues

### CODE REVIEW

File: wallet/src/main/java/ch/admin/bag/covidcertificate/wallet/transfercode/net/DeliveryRepository.kt [hunk 1]
```diff
@@ -11,6 +11,7 @@
 package ch.admin.bag.covidcertificate.wallet.transfercode.net
 
 import android.content.Context
+import android.util.Base64
```
Comment (maurhofer-ubique) [inline wallet/src/main/java/ch/admin/bag/covidcertificate/wallet/transfercode/net/DeliveryRepository.kt:14]: Unused

File: wallet/src/main/java/ch/admin/bag/covidcertificate/wallet/transfercode/worker/TransferWorker.kt [hunk 2]
```diff
@@ -84,7 +84,7 @@ class TransferWorker(context: Context, workerParams: WorkerParameters) : Corouti
 	}
 
 	private suspend fun downloadTransferCertificate(transferCode: TransferCodeModel): TransferState {
-		val keyPair = TransferCodeCrypto.loadKeyPair(transferCode.code)
+		val keyPair = TransferCodeCrypto.loadKeyPair(transferCode.code, context)
```
Comment (maurhofer-ubique) [inline wallet/src/main/java/ch/admin/bag/covidcertificate/wallet/transfercode/worker/TransferWorker.kt:87]: Wouldn't applicationContext work?
(Same on 108)

Then the context wouldn't need to be persisted in a member variable.

**Review by maurhofer-ubique — CHANGES_REQUESTED** [review]

File: wallet/build.gradle [hunk 3]
```diff
@@ -130,6 +129,7 @@ dependencies {
 	implementation 'androidx.viewpager2:viewpager2:1.0.0'
 	implementation 'androidx.security:security-crypto:1.0.0'
 	implementation 'androidx.work:work-runtime-ktx:2.5.0'
+	implementation 'org.bouncycastle:bcprov-jdk15on:1.65'
```
Comment (goebelUB) [inline wallet/build.gradle:132]: Doesn't belong to the androidx block of dependencies

File: wallet/src/main/java/ch/admin/bag/covidcertificate/wallet/transfercode/logic/TransferCodeCrypto.kt [hunk 4]
```diff
@@ -17,21 +17,38 @@ import android.security.keystore.KeyInfo
...
 	const val ANDROID_KEYSTORE_NAME = "AndroidKeyStore"
+	const val BOUNCY_CASTLE_PROVIDER = "BC"
+
+	init {
+		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
+			// Remove platform provided bouncy castle provider and add updated one from library on Android 23 and lower
+			Security.removeProvider(BOUNCY_CASTLE_PROVIDER)
+			Security.addProvider(BouncyCastleProvider())
+		}
+	}
```
Comment (goebelUB) [inline wallet/src/main/java/ch/admin/bag/covidcertificate/wallet/transfercode/logic/TransferCodeCrypto.kt:51]: Do we want to do this "globally" here, or explicitly for all functions? (e.g. `KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_RSA, BOUNCY_CASTLE_PROVIDER)`)

File: wallet/src/main/java/ch/admin/bag/covidcertificate/wallet/data/WalletSecureStorage.kt [hunk 5]
```diff
@@ -39,4 +41,14 @@ class WalletSecureStorage private constructor(context: Context) {
...
+	fun getTransferCodePublicKey(keyAlias: String) = prefs.getString(KEY_TRANSFER_CODE_PUBLIC_KEY_PREFIX + keyAlias, null)
+
+	fun setTransferCodePublicKey(keyAlias: String, encodedKey: String?) =
+		prefs.edit().putString(KEY_TRANSFER_CODE_PUBLIC_KEY_PREFIX + keyAlias, encodedKey).apply()
+
+	fun getTransferCodePrivateKey(keyAlias: String) = prefs.getString(KEY_TRANSFER_CODE_PRIVATE_KEY_PREFIX + keyAlias, null)
+
+	fun setTransferCodePrivateKey(keyAlias: String, encodedKey: String?) =
+		prefs.edit().putString(KEY_TRANSFER_CODE_PRIVATE_KEY_PREFIX + keyAlias, encodedKey).apply()
+
```
Comment (goebelUB) [inline wallet/src/main/java/ch/admin/bag/covidcertificate/wallet/data/WalletSecureStorage.kt:53]: Maybe a comment here warning that these functions are only intended for use on Marshmallow?

**Comment by goebelUB** [comment]
For completeness: with this upgrading the OS from Android 6 to something higher will lose any pending transfer codes. Arguably this case is very rare: Android 7 was released in August 2016 so the chance that anybody upgrades 4 years later is negligible.

File: wallet/src/main/java/ch/admin/bag/covidcertificate/wallet/transfercode/worker/TransferWorker.kt (hunk 2 above)
Comment (M-Wong) [reply wallet/src/main/java/ch/admin/bag/covidcertificate/wallet/transfercode/worker/TransferWorker.kt:87]: True, I didn't see that the applicationContext was available in the worker

File: wallet/src/main/java/ch/admin/bag/covidcertificate/wallet/transfercode/logic/TransferCodeCrypto.kt (hunk 4 above)
Comment (M-Wong) [reply wallet/src/main/java/ch/admin/bag/covidcertificate/wallet/transfercode/logic/TransferCodeCrypto.kt:51]: If you don't add the provider, you can't get it even if you specify it explicitly in the `getInstance()` methods. The `provider` parameter in those methods only specifies the provider you're requesting, it still has to be available in the `Security.getProviders()` list, which by default only contains the outdated BouncyCastle provider from the Android platform

**Comment by M-Wong** [comment]
> For completeness: with this upgrading the OS from Android 6 to something higher will lose any pending transfer codes. Arguably this case is very rare: Android 7 was released in August 2016 so the chance that anybody upgrades 4 years later is negligible.

That is true, however migrating the key pairs from our EncryptedSharedPreferences to the AndroidKeyStore would require us to import those key entries and specifying our own certificate chain in the process. As you said, that case is extremely unlikely and therefore additional resources for developing and testing such a migration seems to me to require more effort than we would gain.

**Review by goebelUB — APPROVED** [review]

**Review by maurhofer-ubique — APPROVED** [review]

**Pull request merged**